main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala

```diff
@@ -0,0 +1,352 @@
+      matches match {
+        // Case 1: No match. Can't reduce the type.
+        case Nil => tpe0
+
+        // Case 2: One match. Use it.
+        case newTpe :: Nil =>
+          tracker.markProgress()
+          newTpe
+
+        // Case 3: Multiple matches. Give back the original type.
+        // TODO CONSTR-SOLVER-2 Right resiliency strategy?
+        case _ :: _ :: _ => tpe0
+      }
+
+    case Type.JvmToType(tpe, loc) =>
+      reduce(tpe) match {
+        case Type.Cst(TypeConstructor.JvmConstructor(constructor), _) =>
+          tracker.markProgress()
+          Type.getFlixType(constructor.getDeclaringClass)
+
+        case Type.Cst(TypeConstructor.JvmMethod(method), _) =>
+          tracker.markProgress()
+          Type.getFlixType(method.getReturnType)
+
+        case Type.Cst(TypeConstructor.JvmField(field), _) =>
+          tracker.markProgress()
+          Type.getFlixType(field.getType)
+
+        case t => Type.JvmToType(t, loc)
+      }
+
+    case Type.JvmToEff(tpe, loc) =>
+      reduce(tpe) match {
+        case Type.Cst(TypeConstructor.JvmConstructor(constructor), _) =>
+          tracker.markProgress()
+          BaseEffects.getConstructorEffs(constructor, loc)
+
+        case Type.Cst(TypeConstructor.JvmMethod(method), _) =>
+          tracker.markProgress()
+          BaseEffects.getMethodEffs(method, loc)
+
+        case t => Type.JvmToType(t, loc)
+      }
+
+    case unresolved@Type.UnresolvedJvmType(member, loc) =>
+      member.map(reduce) match {
+        case JvmMember.JvmConstructor(clazz, tpes) =>
+          lookupConstructor(clazz, tpes) match {
+            case JavaConstructorResolution.Resolved(constructor) =>
+              tracker.markProgress()
+              Type.Cst(TypeConstructor.JvmConstructor(constructor), loc)
+            case _ => unresolved
+          }
+
+        case JvmMember.JvmField(tpe, name) =>
+          lookupField(tpe, name.name) match {
+            case JavaFieldResolution.Resolved(field) =>
+              tracker.markProgress()
+              Type.Cst(TypeConstructor.JvmField(field), loc)
+            case _ => unresolved
+          }
+
+        case JvmMember.JvmMethod(tpe, name, tpes) =>
+          lookupMethod(tpe, name.name, tpes) match {
+            case JavaMethodResolution.Resolved(method) =>
+              tracker.markProgress()
+              Type.Cst(TypeConstructor.JvmMethod(method), loc)
+            case _ => unresolved
+          }
+
+        case JvmMember.JvmStaticMethod(clazz, name, tpes) =>
+          lookupStaticMethod(clazz, name.name, tpes) match {
+            case JavaMethodResolution.Resolved(method) =>
+              tracker.markProgress()
+              Type.Cst(TypeConstructor.JvmMethod(method), loc)
+            case _ => unresolved
+          }
+      }
+  }
+
+  /** Tries to find a constructor of `clazz` that takes arguments of type `ts`. */
+  def lookupConstructor(clazz: Class[?], ts: List[Type]): JavaConstructorResolution = {
```

Comment:
Is the following standard / reused? Do I need to check it?

Reply:
All of the `lookup*` stuff is copy-paste